### ISSUE TITLE
Jasmine not always waiting for empty control flow

### ISSUE THREAD

**Issue by renehamburger** [angular/protractor#3240]
The following code snippet is an example of a situation where the control flow isn't executed completely:

``` js
function log() {
  return browser.controlFlow().execute(() => {
    console.log.apply(console, arguments);
  });
}

it('control flow test', () => {
  const promise = new Promise((resolve) => {
    setTimeout(resolve, 1000);
  });
  log('queue started');
  return promise.then(() => {
    log('queue continues after resolved promise');
    browser.get('http://ng2piano.azurewebsites.net');
    log('queue finished');
    console.log('all commands added to queue');
  });
});
```

The 'queue finished' message is never logged.

Removing the `browser.get()` or returning a webdriver promise of another queued command, e.g. `return browser.sleep(0)`, resolves the issue.

**Comment by juliemr** [comment]
Making sure - this is using Jasmine?

Jasmine does not wait on returned promises, so creating your own native promise and returning it is probably not doing what you think it will. Because the test will finish immediately, the control flow queue is getting added to outside of the test method, which is screwing up our logic.

You could instead use a `done` parameter to delay Jasmine test completion until that is called - that should work. Alternatively, creating your promise via `protractor.promise.Promise` should work.

**Comment by renehamburger** [comment]
Yes, I'm using Protractor 3.3.0 out of the box, so Jasmine/jasminewd2 (which does seem to wait for returned promises in general).

You're right, using the webdriver promise instead of an ES6 (or Q's) promise works. Thanks! Not intuitive, though, so maybe something to add to the docs or FAQs?

But I am still wondering whether all this is pointing to a bug somewhere under the hood, as I couldn't get it to work with `done()` either:
```
it('control flow test', (done) => {
  const promise = new Promise((resolve) => {
    setTimeout(resolve, 1000);
  });
  log('queue started');
  promise.then(() => {
    log('queue continues after resolved promise');
    browser.get('http://ng2piano.azurewebsites.net');
    log('queue finished');
    console.log('all commands added to queue');
    done();
  });
});
```
Same problem as above: all commands within the promise callback are queued but queue is only  executed up to `browser.get()`, which then fails, presumably because the test is finished. And if `browser.get()` is replaced by another command, even `browser.sleep(1000)`, the whole queue is executed for any of the scenarios above.

**Comment by juliemr** [comment]
I'm unable to reproduce with the latest Protractor and your example, so I'll go ahead and close this. Please open a new issue if you have a repro. Thanks!
